### ISSUE TITLE
Race condition in out_loki tenant_id_key (proposed solution inside)

### ISSUE THREAD

**Issue by popaconsta** [fluent/fluent-bit#6911]
## Bug Report

**Describe the bug**
The `tenant_id_key` was introduced to allow dynamic `tenant_id` values within a single running instance of the Loki Output Plugin. In an environment with high concurrency, however, the behaviour becomes unpredictable due to `ctx->dynamic_tenant_id` being written and read concurrently.

**To Reproduce**
Writing a complete description on how to reproduce this bug would be too long and verbose. However, I hope you will find it reasonable that this problem exists when trying to ingest thousands or millions of logs per hour. Looking at the code with the following scenario in mind, should make it pretty obvious:

`Event #1` with `tenant_id=tenant_A` arrives and `ctx->dynamic_tenant_id` is set to `tenant_A`.
`Event #2` with `tenant_id=tenant_B` arrives and `ctx->dynamic_tenant_id` will be overwritten to `tenantB`
Between the arrival of `Event #1` and `Event #2`, the Loki plugin might try to flush `Event #1`. When attempting to read `ctx->dynamic_tenant_id` , its value might be one of the following:

- `tenant_A` (no overwrite yet, this is the lucky case)
- `NULL` (the string buffer was emptied while overwriting the value)
- `tenant_B` (the value was already overwritten before flushing `Event #1`)

**Expected behavior**
`cb_loki_flush ` should read the value from the payload directly and use it as the value of `FLB_LOKI_HEADER_SCOPE ` aka `X-Scope-OrgID` Header(instead of using a shared variable such as `dynamic_tenant_id `). Eventually, it should remove the value if the config requires that label to be removed.

**Additional context**
I have tried to query the logs in Loki and events that are supposed to be under a given `Org ID` (e.g. `Tenant_A`), end up in a different namespace, such as (`Tenant_B`). The wrong value was therefore used for the `X-Scope-OrgID` while flushing the log.

**Comment by leonardo-albertovich** [comment]
Hi @popaconsta, I just saw this issue and verified it myself and it's actually very easy to prove if you modify the code, all you have to do is print the result of `pthread_self()` and the value of `ctx` in line 1083 and then run fluent-bit with this configuration  : 

```
[SERVICE]
    flush 1
    grace 1
    log_level info

[INPUT]
    name dummy
    samples 1000
    dummy {"log": "b", "dog": "maybe"}
    rate 10

[OUTPUT]
    name loki
    match *
    host 127.0.0.1
    port 3000
    tenant_id_key dog
    workers 10
```

The web server doesn't even have to be valid, I just used python with the http.server module and after a very short while I started seeing the intertwined debug lines that proved that different worker threads were using the same context instance and thus modifying the same member concurrently.

I also noticed that `get_tenant_id_from_record` is called from `pack_record` which in turn is called by `loki_compose_payload` for each record in the chunk but there is no way to specify more than one value for `X-Scope-OrgID` which means the last value we find determines the value that will be used for the whole batch which sounds wrong to me (maybe it's just not what happens in most setups but something makes me think we should cut the batch when that value changes, send the request and continue with the next batch of matching IDs)

I think the least intrusive thing we could do right now would be adding an sds output parameter to `loki_compose_payload` that's passed over to `pack_record` and `get_tenant_id_from_record` in a chain so we can properly contain it in the thread and dispose of it after the flush operation completes.

It doesn't sound like the optimal or "correct" approach but it does seem to be the least likely to cause issues or introduce an unexpected default behavior change that could cause issues to existing users.

**Comment by leonardo-albertovich** [comment]
I just sent PR #6931 and would really appreciate it if you could test it and / or review it since I worked on a "vacuum chamber" and don't know enough about loki to be 110% I'm not making something dumb.

I marked the PR as a draft so it's not merged by mistake before someone more qualified reviews it or at least have someone vouch for it.

**Comment by popaconsta** [comment]
Thank you for the quick follow-up @leonardo-albertovich 

I forgot to update the thread, but you are right in your findings: on top of the concurrency problems, records are packed even though they are using different tenant IDs. My guess is that the dynamic tenant id key (i.e. `tenant_id_key`) was NOT thought as a truly dynamic way to express multiple tenant IDs for Loki, but rather as a way to set the key name *once* (most likely at the beginning of the execution) in a "dynamic" way, without altering it later, but at the same time without having it embedded in the configuration (like `tenant_id` does).

I have reviewed your PR and left a comment, so please have a look.

**Comment by EugenFo** [comment]
Hey

Somehow this issue still exists. 
We are also using `tenant_id_key` in the Loki output plugin and trying to split our logs into 2 tenants (separation of infrastructure logs and application logs).
When a lot of logs are being processed then fluent-bit messes the tenant_ids up and some of our infrastructure logs are send to the application tenant.

However when fluent-bit don't get lot of logs then everything works as intended.

We are using fluent-bit version 2.1.5, where the fix from [#6931](https://github.com/fluent/fluent-bit/pull/6931) is included. 

If there is a need to reproduce this issue I will try to create the steps for this.
